### PR TITLE
Fixed get() to propagate events to listeners

### PR DESCRIPTION
**Pull request by maneesht**
Fix for related https://github.com/firebase/firebase-js-sdk/issues/6397

### CODE REVIEW

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java [hunk 1]
```diff
@@ -4548,6 +4548,57 @@ private static FirebaseApp appForDatabaseUrl(String url, String name) {
...
+      FirebaseDatabase writeDb = FirebaseDatabase.getInstance(writeApp);
+      DatabaseReference writeNode = writeDb.getReference().push();
+      DatabaseReference readNode = IntegrationTestHelpers.translateReference(writeNode, readDb);
+      try {
+        Map<String, Object> toWrite =
+            new MapBuilder().put("foo1", "child1").put("foo2", "child2").build();
+        await(writeNode.setValue(toWrite));
+        Task<DataSnapshot> getTask = readNode.get();
+        AtomicBoolean receivedGetValue = new AtomicBoolean();
+        getTask.addOnCompleteListener(
```
Comment (IanWyszynski) [inline firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4576]: Couldn't you ensure that `getTask` is pending by taking the database offline, calling `get()`, register the `ValueEventRegistration`, and then going back online?

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java [hunk 2]
```diff
@@ -4548,6 +4548,57 @@ private static FirebaseApp appForDatabaseUrl(String url, String name) {
...
+        pendingGet = !receivedGetValue.get();
+        DatabaseReference childRef = readNode.child("foo1");
+        new ReadFuture(
+                childRef,
+                events -> {
+                  assertEquals(1, events.size());
+                  assertEquals("child1", events.get(0).getSnapshot().getValue());
+                  return true;
+                })
+            .timedGet();
```
Comment (IanWyszynski) [inline firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4594]: So if you revert this change this call to `timedGet` times out? The test timeout is 70s, maybe we could set it to something a bit lower?  Might require a bit of experimentation.

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java [hunk 3]
```diff
@@ -4548,6 +4548,57 @@ private static FirebaseApp appForDatabaseUrl(String url, String name) {
...
+      throws DatabaseException, InterruptedException, ExecutionException {
+    /**
+     * There's a situation where if you have a pending get() and then an onValue() at the same
+     * location and get resolves before onValue does, onValue gets no listeners
+     */
+    int counter = 0;
+    boolean pendingGet = false;
+    while (counter++ < 5 || !pendingGet) {
+      FirebaseApp writeApp =
+          appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
```
Comment (IanWyszynski) [inline firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4562]: It seems like all we really need in test is the `FirebaseDatabase`. Perhaps we should add a helper:

```
FirebaseDatabase readerDb = IntegrationTestHelpers.newFirebaseDatabase();
```

Just a wrapper over what you have here to make this a bit easier to read.

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java [hunk 4]
```diff
@@ -4548,6 +4548,57 @@ private static FirebaseApp appForDatabaseUrl(String url, String name) {
...
+                DataSnapshot snapshot = task.getResult();
+                assertEquals(toWrite, snapshot.getValue());
+                receivedGetValue.set(true);
+              }
+            });
+        pendingGet = !receivedGetValue.get();
+        DatabaseReference childRef = readNode.child("foo1");
+        new ReadFuture(
+                childRef,
+                events -> {
```
Comment (IanWyszynski) [inline firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4589]: I think this callback will run on the platform event loop, but where does that `getTask.addOnCompleteListener(...)` callback run? I think we have to pass the runloop explicitly.

Comment (IanWyszynski) [reply firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4589]: https://github.com/firebase/firebase-android-sdk/blob/master/firebase-database/src/main/java/com/google/firebase/database/core/Repo.java#L540

Comment (maneesht) [reply firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4589]: That `getTask` is just chaining off of the one in `getValue`, so I thought they would be using the same thread...And `setValue`'s test doesn't seem to specify https://github.com/firebase/firebase-android-sdk/blob/fbbe3e20aaa8d1ef4c4740a02ef8237329ca14a6/firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java#L1592

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java (hunk 2 above)
Comment (maneesht) [reply firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4594]: Ended up changing it to 10s

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java [hunk 5]
```diff
@@ -4548,6 +4548,57 @@ private static FirebaseApp appForDatabaseUrl(String url, String name) {
         name);
   }
 
+  @Test
+  public void testGetPropagatesEventsToListeners() // TODO(mtewani): Convert into smoke test
+      throws DatabaseException, InterruptedException, ExecutionException {
+    /**
+     * There's a situation where if you have a pending get() and then an onValue() at the same
+     * location and get resolves before onValue does, onValue gets no listeners
```
[the remark this answers is not in the document]
Comment (maneesht) [reply firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4556]: Thanks.

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java (hunk 3 above)
Comment (maneesht) [reply firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4562]: Done.

File: firebase-database/src/androidTest/java/com/google/firebase/database/IntegrationTestHelpers.java [hunk 6]
```diff
@@ -342,6 +344,23 @@ public static DataSnapshot referenceAtPath(DatabaseReference node, EventRecord r
...
+        new FirebaseOptions.Builder()
+            .setApplicationId("appid")
+            .setApiKey("apikey")
+            .setDatabaseUrl(url)
+            .build(),
+        name);
+  }
+
+  public static FirebaseDatabase getNewDatabase() {
+    FirebaseApp readApp =
```
Comment (IanWyszynski) [inline firebase-database/src/androidTest/java/com/google/firebase/database/IntegrationTestHelpers.java:359]: nit: We can just call this `app`.

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java [hunk 7]
```diff
@@ -4537,31 +4536,76 @@ public void onComplete(DatabaseError error, DatabaseReference ref) {
...
+  @Test
+  public void testGetPropagatesEventsToListeners()
+      throws DatabaseException, InterruptedException, ExecutionException {
+    /**
+     * There's a situation where if you have a pending `get()` and then an `onValue()` at the same
+     * location and get resolves before onValue does, onValue gets no events. Note: This is a smoke
+     * test due to the fact that it is difficult to reproduce the scenario when the parent get()
+     * request resolves before a child onValue listener.
+     */
+    // To ensure that we don't read the cached results, we need a separate app.
```
Comment (IanWyszynski) [inline firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4548]: This comment can probably be moved down above L4553.

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java [hunk 8]
```diff
@@ -4537,31 +4536,76 @@ public void onComplete(DatabaseError error, DatabaseReference ref) {
...
+    AtomicBoolean successful = new AtomicBoolean();
+    int count = 0;
+    while (count++ < 5 && !successful.get()) {
+
```
Comment (IanWyszynski) [inline firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4552]: nit: remove this empty line (fill it with that comment above).

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java (hunk 4 above)
Comment (IanWyszynski) [reply firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4582]: Ah, you're right.

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java [hunk 9]
```diff
@@ -4537,31 +4536,76 @@ public void onComplete(DatabaseError error, DatabaseReference ref) {
...
+              @Override
+              public void onComplete(@NonNull Task<DataSnapshot> task) {
+                DataSnapshot snapshot = task.getResult();
+                assertEquals(toWrite, snapshot.getValue());
+                resolved.set(true);
+              }
+            });
+
+        DatabaseReference childRef = readNode.child("foo1");
+        readDb.goOnline();
```
Comment (IanWyszynski) [inline firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4575]: Why not move the `goOnline` call to right after we create the `ReadFuture`? That way we don't need to have an explicit thread sleep.

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java [hunk 10]
```diff
@@ -4537,31 +4536,76 @@ public void onComplete(DatabaseError error, DatabaseReference ref) {
...
+     * test due to the fact that it is difficult to reproduce the scenario when the parent get()
+     * request resolves before a child onValue listener.
+     */
+    // To ensure that we don't read the cached results, we need a separate app.
+    AtomicBoolean successful = new AtomicBoolean();
+    int count = 0;
+    while (count++ < 5 && !successful.get()) {
+
+      FirebaseDatabase readDb = IntegrationTestHelpers.getNewDatabase();
+      FirebaseDatabase writeDb = IntegrationTestHelpers.getNewDatabase();
```
Comment (IanWyszynski) [inline firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4554]: We're using the unqualified `getNewDatabase()` elsewhere:

```
FirebaseDatabase writeDb = getNewDatabase();
```

We can be consistent and drop `IntegrationTestHelpers`.

Comment (IanWyszynski) [reply firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4554]: I'd say for all the commonly used helpers in `IntegrationTestHelpers` let's use the unqualified form.

File: firebase-database/src/main/java/com/google/firebase/database/core/Repo.java [hunk 11]
```diff
@@ -574,12 +574,19 @@ public void onCancelled(@NonNull DatabaseError error) {
                             new ValueEventRegistration(repo, listener, spec);
```
Comment (IanWyszynski) [inline firebase-database/src/main/java/com/google/firebase/database/core/Repo.java:574]: Can we use a KeepSyncedEventRegistration for the get query?

https://github.com/firebase/firebase-android-sdk/blob/884e53770ef0fb224b9401ade32a4f8fc4bc4e35/firebase-database/src/main/java/com/google/firebase/database/core/SyncTree.java#L799-L848

File: firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java [hunk 12]
```diff
@@ -4537,31 +4536,76 @@ public void onComplete(DatabaseError error, DatabaseReference ref) {
...
-            .setApplicationId("appid")
-            .setApiKey("apikey")
-            .setDatabaseUrl(url)
-            .build(),
-        name);
+  @Test
+  public void testGetPropagatesEventsToListeners()
+      throws DatabaseException, InterruptedException, ExecutionException {
+    /**
+     * There's a situation where if you have a pending `get()` and then an `onValue()` at the same
```
Comment (maneesht) [inline firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java:4543]: Fix up phrasing to explain exact order of steps.

**Review by IanWyszynski — APPROVED** [review]
